### PR TITLE
ENH: allow broadcasting in discr element

### PR DESCRIPTION
**Pull request by adler-j**
Convenience fix, result should be obvious from tests. Also had to fix an erroneous test in chambolle pock (that was "saved" before this change).

### CODE REVIEW

File: odl/discr/lp_discr.py [hunk 1]
```diff
@@ -286,12 +286,24 @@ def element(self, inp=None, **kwargs):
...
                     raise ValueError(
-                        'input shape {} does not match grid shape {}.'
-                        ''.format(arr.shape, self.shape))
+                        'input shape {} does not match grid shape {}'
+                        ''.format(input_shape, self.shape))
+            elif arr.shape != self.shape and arr.ndim == self.ndim:
+                # Try to broadcast the array if possible
+                for i in range(self.ndim):
+                    if arr.shape[i] == 1 and self.shape[i] != 1:
+                        arr = arr.repeat(self.shape[i], axis=i)
```
Comment (kohr-h) [inline odl/discr/lp_discr.py]: Did you benchmark this? I think this solution is quite slow compared to broadcasting with factor zero. I suggest that you test this and try it against the alternative of building a zero array for the missing dimensions, i.e.
```python
shape = [m if n == 1 and m != 1 else n
         for n, m in zip(arr.shape, self.shape)]
arr = arr + np.zeros(shape)
```

File: odl/space/fspace.py [hunk 2]
```diff
@@ -405,9 +405,14 @@ def __call__(self, x, out=None, **kwargs):
                 out = out.astype(self.out_dtype)
 
             if out_shape != (1,) and out.shape != out_shape:
-                raise ValueError('output shape {} not equal to shape '
-                                 '{} expected from input.'
-                                 ''.format(out.shape, out_shape))
+                # Try to broadcast the returned element if possible
+                for i in range(len(out_shape)):
+                    if out.shape[i] == 1 and out_shape[i] != 1:
+                        out = out.repeat(out_shape[i], axis=i)
```
Comment (kohr-h) [inline odl/space/fspace.py]: same here

**Comment by kohr-h** [comment]
Nice small change, improves usability significantly. Just check which version is faster.

File: odl/discr/lp_discr.py (hunk 1 above)
Comment (kohr-h) [inline odl/discr/lp_discr.py]: CORRECTION:
```python
shape = [m if n == 1 and m != 1 else 1
         for n, m in zip(arr.shape, self.shape)]
arr = arr + np.zeros(shape)
```

**Comment by adler-j** [comment]
Results from performance tests:

code 1
```python
shape = [m if n == 1 and m != 1 else 1
         for n, m in zip(out.shape, out_shape)]
out = out + np.zeros(shape, dtype=out.dtype)
```

code 2
```python
shape = [m if n == 1 and m != 1 else 1
         for n, m in zip(out.shape, out_shape)]
out = np.tile(out, shape)
```

code 3
```python
for i in range(len(out_shape)):
    if out.shape[i] == 1 and out_shape[i] != 1:
        out = out.repeat(out_shape[i], axis=i)
```

code 4
```python
out = np.broadcast_to(out, out_shape)
```


Results with broadcast (1, 1, 500) -> (500, 500, 500)

| Case |  Timing |
|------|---------|
|  1   |  0.084s |
|  2   |  0.404s  |
|  3   |  0.410s  |
|  4   |  0.084s |

Results with broadcast (1, 500, 500) -> (500, 500, 500)

| Case |  Timing |
|------|---------|
|  1   |  0.151s  |
|  2   |  0.396s  |
|  3   |  0.420s  |
|  4    |  0.086s |

The winner is clearly `broadcast_to` followed by add zeros. The issue with `broadcast_to` is that it needs numpy > 1.10, we currently require numpy 1.9. I'll make some kind of if statement to check that and fallback to add zeros.

**Comment by adler-j** [comment]
Updated according to the above, please review again.

File: odl/discr/lp_discr.py [hunk 3]
```diff
@@ -286,12 +285,24 @@ def element(self, inp=None, **kwargs):
...
-                        'input shape {} does not match grid shape {}.'
-                        ''.format(arr.shape, self.shape))
+                        'input shape {} does not match grid shape {}'
+                        ''.format(input_shape, self.shape))
+            elif arr.shape != self.shape and arr.ndim == self.ndim:
+                # Try to broadcast the array if possible
+                try:
+                    arr = np.broadcast_to(arr, self.shape)
+                except AttributeError:
+                    # The above requires numpy 1.9, fallback impl else
```
Comment (kohr-h) [inline odl/discr/lp_discr.py]: 1.9 or 1.10?

Comment (adler-j) [inline odl/discr/lp_discr.py]: 1.10

**Comment by kohr-h** [comment]
Otherwise good!

**Pull request merged**